### PR TITLE
Add new Python and C++/CUDA Custom Op tutorials

### PR DESCRIPTION
**Pull request by zou3519**
I want to land this before PyTorch 2.4 (so we can link to these in PyTorch's nightly documentation) and then have a follow-up PR for 2.4 that actually runs the scripts (so that they can generate outputs).

https://github.com/pytorch/pytorch/issues/127443 to remind myself of the above.

NB: These two tutorials replace all of the existing custom ops (and cpp extensions) tutorials:
- advanced/cpp_extension
- advanced/torch_script_custom_ops
- advanced/torch_script_custom_classes
- advanced/dispatcher

In a follow-up PR we will add warnings to all of those tutorials stating that they are deprecated but we will preserve the text just in case people still need them (e.g. if they are not using PyTorch 2.4).

Test Plan:
- I tested these locally.

Fixes #ISSUE_NUMBER

## Checklist

- [ ] The issue that is being fixed is referred in the description (see above "Fixes #ISSUE_NUMBER")
- [ ] Only one issue is addressed in this pull request
- [ ] Labels from the issue that this PR is fixing are added to this pull request
- [ ] No unnecessary issues are included into this pull request.

### CODE REVIEW

**Comment by zou3519** [comment]
@svekars is there a way to bypass the pyspelling check?

File: advanced_source/cpp_custom_ops.rst [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+  def mymuladd(a: Tensor, b: Tensor, c: float):
+      return a * b + c
+
+You can find the end-to-end working example for this tutorial over at
+https://github.com/pytorch/extension-cpp .
+
+Build System
+------------
+
+If you author custom C++/CUDA code, it needs to be compiled somehow.
```
Comment (svekars) [inline advanced_source/cpp_custom_ops.rst:26]:
```suggestion
If you are developing custom C++/CUDA code, it must be compiled.
```

File: advanced_source/cpp_custom_ops.rst [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+Note that if you’re interfacing with a Python library that already has bindings
+to precompiled C++/CUDA code, then you may actually want to write a Python custom operator
```
Comment (svekars) [inline advanced_source/cpp_custom_ops.rst:28]:
```suggestion
to precompiled C++/CUDA code, you might consider writing a customa Python operator instead.
```

File: index.rst [hunk 3]
```diff
@@ -416,6 +416,27 @@ Welcome to PyTorch Tutorials
...
+
+.. customcarditem::
+   :header: Custom C++ and CUDA Operators
+   :card_description: How to extend PyTorch with custom C++ and CUDA operators.
+   :image: _static/img/thumbnails/cropped/Custom-Cpp-and-CUDA-Extensions.png
+   :link: advanced/cpp_custom_ops.html
+   :tags: Extending-PyTorch,Frontend-APIs,C++,CUDA
+
+.. customcarditem::
+   :header: Extending TorchScript with Custom C++ Operators
```
Comment (albanD) [inline index.rst:434]: Shouldn't this one just be removed?

File: advanced_source/python_custom_ops.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+.. note::
+   This tutorial is for PyTorch 2.4+ and the PyTorch nightlies.
+
+PyTorch offers a large library of operators that work on Tensors (e.g.
+torch.add, torch.sum, etc). However, you may wish to use a new customized
+operator with PyTorch, perhaps written by a third-party library. This tutorial
+shows how to wrap Python functions so that they behave like PyTorch native
+operators. Reasons why you may wish to create a custom op in PyTorch include:
+
+- Black-box-ing an arbitrary Python function for use with torch.compile
```
Comment (albanD) [inline advanced_source/python_custom_ops.py:16]: Blackboxing needs defining here

File: advanced_source/python_custom_ops.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+- Adding training support to an arbitrary Python function
+
+Please note that if your operation can be expressed as a composition of
+existing PyTorch ops, then there is usually no need to use the custom op
+API -- everything (e.g. torch.compile, training support) should just work.
+"""
+######################################################################
+# Wrapping PIL's crop into a custom op
```
Comment (albanD) [inline advanced_source/python_custom_ops.py:24]:
```suggestion
# Example: Wrapping PIL's crop into a custom op
```

File: advanced_source/python_custom_ops.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+
+
+img = torch.ones(3, 64, 64)
+img *= torch.linspace(0, 1, steps=64) * torch.linspace(0, 1, steps=64).unsqueeze(-1)
+display(img)
+cropped_img = crop(img, (10, 10, 50, 50))
+display(cropped_img)
+
+######################################################################
+# ``crop`` doesn't work performantly out-of-the-box with torch.compile. The
```
Comment (albanD) [inline advanced_source/python_custom_ops.py:50]:
```suggestion
# ``crop`` is not handled effectively out-of-the-box by torch.compile. The
```

File: advanced_source/python_custom_ops.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+# following code leads to an error when run.
+
+"""
```
Comment (albanD) [inline advanced_source/python_custom_ops.py:53]: Why is this one not code?

File: advanced_source/python_custom_ops.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+@torch.compile(fullgraph=True)
```
Comment (albanD) [inline advanced_source/python_custom_ops.py:54]: I guess "effectively handled" above means no graph break? If so, we should state it clearly. Maybe linking to some compile doc explaining why graph breaks are not desireable

File: advanced_source/python_custom_ops.py [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+
+cropped_img = f(img)
+"""
+
+######################################################################
+# In order to black-box ``crop`` for use with ``torch.compile``, we need to do two things:
+#
+# - wrap the function into a PyTorch custom op.
+# - add a "FakeTensor kernel" (aka "meta kernel") to the op. Given the metadata (e.g. shapes)
+# of the input Tensors, this function says how to compute the metadata of the output Tensor(s).
```
Comment (albanD) [inline advanced_source/python_custom_ops.py:66]: Rendering of this is wrong

File: advanced_source/python_custom_ops.py [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+
+# Use register_fake to add a FakeTensor kernel for the op
+@crop.register_fake
+def _(pic, box):
+    channels = pic.shape[0]
+    x0, y0, x1, y1 = box
+    return pic.new_empty(channels, y1 - y0, x1 - x0)
+
+######################################################################
+# After this, crop now works with torch.compile:
```
Comment (albanD) [inline advanced_source/python_custom_ops.py:88]: It already worked. Only works without graph break now

File: advanced_source/python_custom_ops.py [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+cropped_img = f(img)
+display(img)
+display(cropped_img)
+
+######################################################################
+# Adding training support for crop
+# --------------------------------
+# Use ``torch.library.register_autograd`` to add training support for an operator.
+# Prefer this over directly using ``torch.autograd.Function``; some compositions of
+# ``autograd.Function`` with PyTorch operator registration APIs can lead to (and
```
Comment (albanD) [inline advanced_source/python_custom_ops.py:103]: Is that in the context of torch.compile?

File: advanced_source/python_custom_ops.py [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+######################################################################
+# This is the correct gradient, with 1s (white) in the cropped region and 0s
+# (black) in the unused region.
+
+######################################################################
+# Testing Python Custom Ops
+# -------------------------
+# Use torch.library.opcheck to test that the custom op was registered
+# correctly. This does not test that the gradients are mathematically correct;
+# please write separate tests for that (either manual ones or torch.autograd.gradcheck).
```
Comment (albanD) [inline advanced_source/python_custom_ops.py:159]: We should really just recommend gradcheck

File: advanced_source/python_custom_ops.py [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+#
+# To use opcheck, pass it a set of example inputs to test against. If your
+# operator supports training, then the examples should include Tensors that
+# require grad. If your operator supports multiple devices, then the examplesxi
```
Comment (albanD) [inline advanced_source/python_custom_ops.py:163]:
```suggestion
# require grad. If your operator supports multiple devices, then the examples
```

File: advanced_source/cpp_custom_ops.rst [hunk 14]
```diff
@@ -0,0 +1,365 @@
...
+
+Please do not return any mutated Tensors as outputs of the operator; this will
+run you into problems later down the line.
+
+Conclusion
+----------
+In this tutorial, we went over the recommended approach to integrating Custom C++
+and CUDA operators with PyTorch. The TORCH_LIBRARY/torch.library APIs are fairly
+low-level; more detail about how to use them can be found over at
+`The Custom Operators Manual <https://pytorch.org/docs/main/notes/custom_operators.html>`_
```
Comment (svekars) [inline advanced_source/cpp_custom_ops.rst:363]:
```suggestion
`The Custom Operators Manual <https://pytorch.org/docs/main/notes/custom_operators.html>`_.
```

File: advanced_source/cpp_custom_ops.rst [hunk 15]
```diff
@@ -0,0 +1,365 @@
...
+known as a “meta kernel” or “abstract impl”). FakeTensors are Tensors that have
+metadata (i.e. shape, dtype, device) but no data: the FakeTensor kernel for an
+operator specifies how to compute the metadata of output tensors given the metadata of input tensors.
+
+We recommend that this be done from Python via the `torch.library.register_fake` API,
+though it is possible to do this from C++ as well (see
+`The Custom Operators Manual <https://pytorch.org/docs/main/notes/custom_operators.html>`_
+for more details).
+
+.. code-block:: python
```
[the remark this answers is not in the document]
Comment (zou3519) [reply advanced_source/cpp_custom_ops.rst:160]: I need to figure out what orders are permissable

File: advanced_source/cpp_custom_ops.rst [hunk 16]
```diff
@@ -0,0 +1,365 @@
...
+  torch.library.register_autograd(
+      "extension_cpp::mymuladd", _backward, setup_context=_setup_context)
+
+How to test an operator
+-----------------------
+Use torch.library.opcheck to test that the custom op was registered correctly.
+This does not test that the gradients are mathematically correct; please write
+separate tests for that (either manual ones or torch.autograd.gradcheck).
+
+.. code-block:: python
```
[the remark this answers is not in the document]
Comment (zou3519) [reply advanced_source/cpp_custom_ops.rst:276]: I wrote them at different times. I can make them the same if we want.

File: advanced_source/python_custom_ops.py (hunk 12 above)
Comment (zou3519) [reply advanced_source/python_custom_ops.py:161]: sometimes people don't write custom ops for torch.double, but gradcheck doesn't seem to work well with torch.float

**Comment by svekars** [comment]
> @svekars is there a way to bypass the pyspelling check?

@zou3519 Please fix what's fixable. Words like PIL's you can add to [en-wordlist.txt](https://github.com/pytorch/tutorials/blob/main/en-wordlist.txt). Things like FakeTensor should be put in backticks (``FakeTensor``) as a class (intersphinx should work too) or you could write *fake tensor* or *Fake Tensor*.

**Comment by zou3519** [comment]
@svekars, @albanD -- thank you for the detailed reviews and suggestions! This should be ready for another pass through now.

File: advanced_source/cpp_custom_ops.rst (hunk 15 above)
[the remark this answers is not in the document]
Comment (zou3519) [reply advanced_source/cpp_custom_ops.rst:168]: Updated.

File: advanced_source/cpp_custom_ops.rst [hunk 17]
```diff
@@ -0,0 +1,405 @@
...
+and CUDA operator that composes with PyTorch subsystems. The semantics of
+the operation are as follows:
+
+.. code-block:: python
+
+  def mymuladd(a: Tensor, b: Tensor, c: float):
+      return a * b + c
+
+You can find the end-to-end working example for this tutorial over at
+https://github.com/pytorch/extension-cpp .
```
Comment (svekars) [inline advanced_source/cpp_custom_ops.rst:21]:
```suggestion
You can find the end-to-end working example for this tutorial `here <https://github.com/pytorch/extension-cpp>`__.
```

File: advanced_source/cpp_custom_ops.rst [hunk 18]
```diff
@@ -0,0 +1,405 @@
...
+
```
Comment (svekars) [inline advanced_source/cpp_custom_ops.rst:22]: Can we add What you will learn  and prerequisites like [here](https://github.com/pytorch/tutorials/blob/main/beginner_source/template_tutorial.py#L9)?
Example:
What you will learn
* How to author and integrate custom operators written in C++ 
* How to compile custom C++/CUDA code and registering it within the PyTorch framework to make it accessible from Python.
* Advanced integration techniques

Prerequisites
* PyTorch 2.4 or later
* Read `torch.utils.cpp_extension <https://pytorch.org/docs/stable/cpp_extension.html>`_
* Basic understanding of C++ and CUDA programming

File: advanced_source/python_custom_ops.py [hunk 19]
```diff
@@ -0,0 +1,254 @@
...
+  into the function).
+- Adding training support to an arbitrary Python function
+
+Please note that if your operation can be expressed as a composition of
+existing PyTorch operators, then there is usually no need to use the custom operator
+API -- everything (for example ``torch.compile``, training support) should
+just work.
+"""
+######################################################################
+# Example: Wrapping PIL's crop into a custom operator
```
Comment (svekars) [inline advanced_source/python_custom_ops.py:29]: Sam as above, add two [cards](https://github.com/pytorch/tutorials/blob/main/beginner_source/template_tutorial.py#L9): what you will learn and prerequisites. Example
What you will learn
* How to create and integrate custom Python functions as PyTorch operators.
* How to optimize with torch.compile and support training with autograd.
* How to add ``FakeTensor`` kernels, register autograd support, and validate operators with ``torch.library.opcheck``.

Prerequisites
* PyTorch 2.4 or later
* Read `torch.compile <link>`__
* Complete `link to relevant tutorial`

**Review by svekars — COMMENTED** [review]
Looks good overall, asking to add what you will learn and prerequisites.

File: advanced_source/cpp_custom_ops.rst [hunk 20]
```diff
@@ -0,0 +1,405 @@
...
+and then call the ``torch.library`` registration APIs. This can happen in one
+of two ways:
+
+1. If you're following this tutorial, importing the Python C extension module
+   we created will load the C++ custom operator definitions.
+2. If your C++ custom operator is located in a shared library object, you can
+   also use ``torch.ops.load_library("/path/to/library.so")`` to load it.
+
+    
+How to add training (autograd) support for an operator
```
Comment (svekars) [inline advanced_source/cpp_custom_ops.rst:203]:
```suggestion
Adding training (autograd) support for an operator
```

File: advanced_source/cpp_custom_ops.rst [hunk 21]
```diff
@@ -0,0 +1,405 @@
...
+      result = torch.ops.extension_cpp.mymuladd(*args)
+      expected = reference_muladd(*args)
+      torch.testing.assert_close(result, expected)
+  
+      # Use opcheck to check for incorrect usage of operator registration APIs
+      torch.library.opcheck(torch.ops.extension_cpp.mymuladd.default, args)
+
+.. _mutable-ops:
+
+How to create mutable operators
```
Comment (svekars) [inline advanced_source/cpp_custom_ops.rst:345]:
```suggestion
Creating mutable operators
```

File: advanced_source/custom_ops_landing_page.rst [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+``torch.sum``, etc). However, you may wish to bring a new custom operation to PyTorch
+and get it to work with subsystems like ``torch.compile``, autograd, and ``torch.vmap``.
+In order to do so, you must register the custom operation with PyTorch via the Python
+`torch.library docs <https://pytorch.org/docs/stable/library.html>`_ or C++ ``TORCH_LIBRARY``
+APIs.
+
+TL;DR
+-----
+
+How do I author a custom op from Python?
```
Comment (svekars) [inline advanced_source/custom_ops_landing_page.rst:16]:
```suggestion
Creating a custom operator from Python?
```

File: advanced_source/custom_ops_landing_page.rst [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+
+Please see :ref:`python-custom-ops-tutorial`.
+
+You may wish to author a custom operator from Python (as opposed to C++) if:
+- you have a Python function you want PyTorch to treat as an opaque callable, especially with
+respect to ``torch.compile`` and ``torch.export``.
+- you have some Python bindings to C++/CUDA kernels and want those to compose with PyTorch
+subsystems (like ``torch.compile`` or ``torch.autograd``)
+
+How do I integrate custom C++ and/or CUDA code with PyTorch?
```
Comment (svekars) [inline advanced_source/custom_ops_landing_page.rst:27]:
```suggestion
Integrating custom C++ and/or CUDA code with PyTorch
```

File: advanced_source/cpp_custom_ops.rst [hunk 24]
```diff
@@ -0,0 +1,405 @@
...
+---------------------------------------------------------
+First, let’s write a C++ function that computes ``mymuladd``:
+
+.. code-block:: cpp
+
+   at::Tensor mymuladd_cpu(at::Tensor a, const at::Tensor& b, double c) {
+     TORCH_CHECK(a.sizes() == b.sizes());
+     TORCH_CHECK(a.dtype() == at::kFloat);
+     TORCH_CHECK(b.dtype() == at::kFloat);
+     TORCH_INTERNAL_ASSERT(a.device().type() == at::DeviceType::CPU);
```
Comment (albanD) [inline advanced_source/cpp_custom_ops.rst:65]: nit: internal assert in user kernel is weird?

File: index.rst [hunk 25]
```diff
@@ -1067,6 +1081,8 @@ Additional Resources
    :hidden:
    :caption: Extending PyTorch
 
+   advanced/python_custom_ops
```
Comment (albanD) [inline index.rst:1084]: I would add "PyTorch Custom Operators Landing Page" to the "Introduction to PyTorch" list.

**Review by albanD — APPROVED** [review]
SGTM
